# Building a simple chatbot with Retrieval Augmented Generation (RAG) using LlamaIndex

![Retrieval Augmented Generation(RAG)](https://miro.medium.com/v2/resize:fit:720/format:webp/1*UyhiO87T-hejRhqI7EwvgA.png)


## Import libraries, API and set filepath

In [1]:
%pip install llama-index==0.8.12 pypdf sentence-transformers ragas openai
import os
os.environ['OPENAI_API_KEY'] = "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" # replace with your API key

from llama_index import Document, GPTVectorStoreIndex, ServiceContext
from llama_index.readers import BeautifulSoupWebReader, SimpleDirectoryReader
from llama_index.llms import OpenAI
from llama_index.evaluation import DatasetGenerator

import openai

Note: you may need to restart the kernel to use updated packages.


In [14]:
# set filepath to my data directory 

current_dir = os.getcwd()
data_dir = os.path.join(current_dir, "data")

## Load the data

According to [LlamaIndex's documentation](https://gpt-index.readthedocs.io/en/latest/examples/data_connectors/simple_directory_reader.html), the `SimpleDirectoryReader` is the most commonly used data connector that just works. Simply pass in a input directory or a list of files. It will select the best file reader based on the file extensions. 

In this use case here, there are PDFs and html pages from different sources, articles, books and summaries provided on the rich tapestry of cultures in Singapore. A snapshot has been uploaded in LLM model in order to provide greater context for querying.

In [15]:
filename_fn = lambda filename: {'file_name': filename}
# SDR is most common reader of dictionaries, can even use LLoader which reads from online.
pdfhtml_docs = SimpleDirectoryReader(input_dir=data_dir, exclude_hidden=True, file_metadata=filename_fn).load_data()
print([x.doc_id for x in pdfhtml_docs])
print(f"Loaded {len(pdfhtml_docs)} docs")

['307cb60f-24e2-499e-9385-9f1411a18b75', '624c2c61-a5ba-4b6c-b6d2-c12dd31e8831', '9d3d3684-a60d-46aa-86b9-17911e6f6774', 'da65a452-84f0-49db-ad34-a6a02e9b8029', '0d643263-0152-4418-9e78-a677375db9ef', '6cbe6640-820d-41a7-83fd-be586b84c650', '543b3a23-a993-4e15-9907-1f279fa818b0', '18bb5164-086b-4528-8c03-7897513b963e', '4b0249b9-a676-416c-8506-4f7084292048', 'd85fe66c-b5d8-4bdb-8b24-a91ef3441f7e', 'cb59caa3-bb65-4eca-86a0-50bd20e5c929', '2ac98306-e211-4348-9db2-6b9b94ab9528', 'b3ffd7ba-f185-46b0-8f5d-7f64edcf46af', 'aeb4ee83-16b0-4fbc-9bec-736c1935e044', '3ae0eee5-218f-468f-b330-9a6180aa2512', '534fc343-8192-4f7e-90c7-60cd0434ad70', 'a77f3676-853f-4b4d-80bd-4bdb660dbc3c', '78a4df2b-7957-41e8-bc4b-c9d36bf1d753', '1db2df99-832c-4193-a2dc-9db0c50e0a7c', 'c23084a0-56af-4c89-9a0a-fb8304ac6328', '7f72101c-a4b9-4427-849d-588dbd878aba', '523627a1-3ba3-4444-b5a4-0b514d576cc6', 'a1f922dd-0734-4017-9bd8-015d9b33f4d5', '77e769fd-81bd-48df-aed5-359a3601c638', '467e38c7-4ebe-4338-93fe-e029c866cf22',

## Build index

With all the data loaded, we can construct the index for the chatbot. There are 4 types of indexing: Summary index, VectorStore Index, Tree Index and Keyword Table Index. Here we are using VectorStore Index, which is also one of the most common types of indexing.

In [20]:
openai.api_key = os.getenv("OPENAI_API_KEY")

# for more info on service context, refer to 
# https://gpt-index.readthedocs.io/en/latest/core_modules/supporting_modules/service_context.html
service_context = ServiceContext.from_defaults(
    llm=OpenAI(model="gpt-3.5-turbo", temperature=0) # degree of randomness from 0 to 1. 
)
docs = pdfhtml_docs 

#VectorStotre index to get embeddings, it creates a store of vectors.
index = GPTVectorStoreIndex.from_documents(documents=docs, service_context=service_context)

In [21]:
#https://medium.com/cyberark-engineering/how-to-add-context-to-openai-gpt-with-llama-index-1c33c6a44055
index.storage_context.persist("./data/index.vecstore")

## Train generation

In [21]:
# Shuffle the documents
import random

random.seed(42)
random.shuffle(docs)

gpt_context = ServiceContext.from_defaults(
    llm=OpenAI(model="gpt-3.5-turbo", temperature=0)
)

In [22]:
question_gen_query = (
    "You are working as a war correspondent focusing on education and updating. \
    Your task is to answer queries from the public. \
    Using the provided context from documents on Israel-Palestine conflict, \
    formulate a single question that captures an important fact from the context. \
    Restrict the question to the context information provided."
)

# find out more about question generation from 
# https://gpt-index.readthedocs.io/en/latest/examples/evaluation/QuestionGeneration.html

dataset_generator = DatasetGenerator.from_documents(
    docs[:50],
    question_gen_query=question_gen_query,
    service_context=gpt_context,
)

In [23]:
questions = dataset_generator.generate_questions_from_nodes(num=40)
print("Generated ", len(questions), " questions")

Generated  40  questions


In [24]:
with open("train_questions.txt", "w") as f:
    for question in questions:
        f.write(question + "\n")

## Eval generation

In [25]:
dataset_generator = DatasetGenerator.from_documents(
    docs[
        50:
    ],  # since we generated ~1 question for 40 documents, we can skip the first 40
    question_gen_query=question_gen_query,
    service_context=gpt_context,
)

In [26]:
questions = dataset_generator.generate_questions_from_nodes(num=40)
print("Generated ", len(questions), " questions")

Generated  40  questions


In [27]:
with open("eval_questions.txt", "w") as f:
    for question in questions:
        f.write(question + "\n")

## Initial Eval with GPT-3.5-Turbo Query Engine

For this eval, we will be using the [`ragas` evaluation library](https://github.com/explodinggradients/ragas).

Ragas has a ton of evaluation metrics for RAG pipelines, and you can read about them [here](https://github.com/explodinggradients/ragas/blob/main/docs/metrics.md).

For this notebook, we will be using the following two metrics

- `answer_relevancy` - This measures how relevant is the generated answer to the prompt. If the generated answer is incomplete or contains redundant information the score will be low. This is quantified by working out the chance of an LLM generating the given question using the generated answer. Values range (0,1), higher the better.
- `faithfulness` - This measures the factual consistency of the generated answer against the given context. This is done using a multi step paradigm that includes creation of statements from the generated answer followed by verifying each of these statements against the context. The answer is scaled to (0,1) range. Higher the better.

In [28]:
questions = []
with open("eval_questions.txt", "r") as f:
    for line in f:
        questions.append(line.strip())

In [29]:
from llama_index import VectorStoreIndex

# limit the context window to 2048 tokens so that refine is used
gpt_context = ServiceContext.from_defaults(
    llm=OpenAI(model="gpt-3.5-turbo", temperature=0), context_window=2048
)

index = VectorStoreIndex.from_documents(docs, service_context=gpt_context)

query_engine = index.as_query_engine(similarity_top_k=2)

In [30]:
contexts = []
answers = []

for question in questions:
    response = query_engine.query(question)
    contexts.append([x.node.get_content() for x in response.source_nodes])
    answers.append(str(response))

In [31]:
from datasets import Dataset
from ragas import evaluate
from ragas.metrics import answer_relevancy, faithfulness

ds = Dataset.from_dict(
    {
        "question": questions,
        "answer": answers,
        "contexts": contexts,
    }
)

result = evaluate(ds, [answer_relevancy, faithfulness])
print(result)

C:\Users\rusta\anaconda3\Lib\site-packages\huggingface_hub\file_download.py:137: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\rusta\.cache\huggingface\hub. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to see activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


evaluating with [answer_relevancy]


100%|████████████████████████████████████████████████████████████████████████████████████| 3/3 [03:59<00:00, 79.88s/it]


evaluating with [faithfulness]


100%|███████████████████████████████████████████████████████████████████████████████████| 3/3 [17:39<00:00, 353.16s/it]


{'ragas_score': 0.9312, 'answer_relevancy': 0.9729, 'faithfulness': 0.8929}


## GPT-4 to Generate Training Data

Here, we use GPT-4 and the `OpenAIFineTuningHandler` to collect data that we want to train on.

In [33]:
from llama_index import ServiceContext
from llama_index.llms import OpenAI
from llama_index.callbacks import OpenAIFineTuningHandler
from llama_index.callbacks import CallbackManager

finetuning_handler = OpenAIFineTuningHandler()
callback_manager = CallbackManager([finetuning_handler])

gpt_4_context = ServiceContext.from_defaults(
    llm=OpenAI(model="gpt-4", temperature=0),
    context_window=2048,  # limit the context window artifically to test refine process
    callback_manager=callback_manager,
)

In [40]:
questions = []
with open("train_questions.txt", "r") as f:
    for line in f:
        questions.append(line.strip())

In [41]:
from llama_index import VectorStoreIndex

index = VectorStoreIndex.from_documents(docs, service_context=gpt_4_context)

query_engine = index.as_query_engine(similarity_top_k=2)

In [42]:
for question in questions:
    response = query_engine.query(question)

## Create `OpenAIFinetuneEngine`

We create an `OpenAIFinetuneEngine`: the finetune engine will take care of launching a finetuning job, and returning an LLM model that you can directly plugin to the rest of LlamaIndex workflows.

We use the default constructor, but we can also directly pass in our finetuning_handler into this engine with the `from_finetuning_handler` class method.



In [59]:
finetuning_handler.save_finetuning_events("finetuning_events.jsonl")

Wrote 62 examples to finetuning_events.jsonl


In [60]:
from llama_index.finetuning import OpenAIFinetuneEngine

finetune_engine = OpenAIFinetuneEngine(
    "gpt-3.5-turbo",
    "finetuning_events.jsonl",
    # start_job_id="<start-job-id>"  # if you have an existing job, can specify id here
)

# finetune_engine = OpenAIFinetuneEngine.from_finetuning_handler(
#     finetuning_handler,
#     "gpt-3.5-turbo",
#     "tmp.jsonl"
# )

In [70]:
finetune_engine.finetune()

Num examples: 62
First example:
{'role': 'system', 'content': "You are an expert Q&A system that is trusted around the world.\nAlways answer the query using the provided context information, and not prior knowledge.\nSome rules to follow:\n1. Never directly reference the given context in your answer.\n2. Avoid statements like 'Based on the context, ...' or 'The context information ...' or anything along those lines."}
{'role': 'user', 'content': 'Context information is below.\n---------------------\npage_label: 10\nfile_name: C:\\Users\\rusta\\Downloads\\NGbot-main\\NGbot-main\\data\\History of Israel Palestine.pdf\n\nISRAELI-PALESTINIAN CONFLICT \nIn 2002, Israel starting building a barrier around \nthe West Bank which actually cuts into the West \nBank.  The barrier, which is not yet   complete, \nhas already affected the lives of hundreds of thou-\nsands of Palestinians. The barrier encircles Pales-\ntinian towns and villages , separating communities \nand families from each other, 

![notification of successful finetuned job](./screenshots/ftjobnotif.png)

In [95]:
finetune_engine.get_current_job()

<FineTuningJob fine_tuning.job id=ftjob-IisQa3ctIwF2P1zwn1NsC6gG at 0x15af5df48f0> JSON: {
  "object": "fine_tuning.job",
  "id": "ftjob-IisQa3ctIwF2P1zwn1NsC6gG",
  "model": "gpt-3.5-turbo-0613",
  "created_at": 1698321466,
  "finished_at": 1698321995,
  "fine_tuned_model": "ft:gpt-3.5-turbo-0613:personal::8DtQDU3k",
  "organization_id": "org-VbFwFDxSNuZFvXxi8u4oxJN7",
  "result_files": [
    "file-u07DUVL3WrdZJAj4H77n3aIh"
  ],
  "status": "succeeded",
  "validation_file": null,
  "training_file": "file-4uvtIhUDu2wo82D46cXCcr76",
  "hyperparameters": {
    "n_epochs": 3
  },
  "trained_tokens": 219552,
  "error": null
}

In [84]:
#Evaluates job status
job = finetune_engine.get_current_job()
print(f"Job Status: {job.status}")

#Different states: Failed, Validating Files, Running, Succeeded

Job Status: succeeded


In [85]:
ft_llm = finetune_engine.get_finetuned_model(temperature=0)

## Evaluation

After some time, your model will be done training!

The next step is running our fine-tuned model on our eval dataset again to measure any performance increase.

In [86]:
from llama_index import ServiceContext
from llama_index.llms import OpenAI
from llama_index.callbacks import OpenAIFineTuningHandler
from llama_index.callbacks import CallbackManager


# Option 1: pass in ft_llm directly into ServiceContext
ft_context = ServiceContext.from_defaults(
    llm=ft_llm,
    context_window=2048,  # limit the context window artifically to test refine process
)

# # Option 2: you can also specify the model name manually
# ft_model_name = "ft:gpt-3.5-turbo-0613:..."
# ft_context = ServiceContext.from_defaults(
#     llm=OpenAI(model=ft_model_name, temperature=0.3),
#     context_window=2048,  # limit the context window artifically to test refine process
# )

In [30]:
questions = []
with open("eval_questions.txt", "r") as f:
    for line in f:
        questions.append(line.strip())

In [39]:
from llama_index import VectorStoreIndex

index = VectorStoreIndex.from_documents(docs, service_context=ft_context)

query_engine = index.as_query_engine(similarity_top_k=2)

NameError: name 'ft_context' is not defined

In [34]:
contexts = []
answers = []

for question in questions:
    response = query_engine.query(question)
    contexts.append([x.node.get_content() for x in response.source_nodes])
    answers.append(str(response))

In [35]:
from datasets import Dataset
from ragas import evaluate
from ragas.metrics import answer_relevancy, faithfulness

ds = Dataset.from_dict(
    {
        "question": questions,
        "answer": answers,
        "contexts": contexts,
    }
)

result = evaluate(ds, [answer_relevancy, faithfulness])
print(result)

evaluating with [answer_relevancy]


100%|████████████████████████████████████████████████████| 3/3 [01:24<00:00, 28.02s/it]


evaluating with [faithfulness]


100%|████████████████████████████████████████████████████| 3/3 [04:34<00:00, 91.43s/it]


{'ragas_score': 0.8086, 'answer_relevancy': 0.9284, 'faithfulness': 0.7162}


| Model            | RAGAS Score | Answer Relevancy | Faithfulness |
|------------------|-------------|------------------|--------------|
| GPT-3.5-Turbo    | 0.7025      | 0.9459           | 0.5588       |
| Finetuned        | 0.8086      | 0.9284           | 0.7162       |

